README.md
# 然之协同系统(ranzhi)

> 适配 好雨云 **一键部署** ，源码与官方同步更新。感谢 [易软天创](http://www.cnezsoft.com/) 为我们带来如此优秀的开源软件！


<img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101450598729.png" width="145px" height="50px"></img>

然之协同系统由客户管理(crm)、日常办公(oa)、现金记账(cash)、团队分享(team)和应用导航(ips)五大模块组成，主要面向中小团队的企业内部管理。和市面上其他的产品相比，然之协同更专注于提供一体化、精简的解决方案。

更多信息参见[官方介绍](https://www.ranzhico.com/book/ranzhi.html)

<a href="#" target="_blank" ><img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101455383121.png" width="147" height="32"></img></a>

# 目录
- [部署到好雨云](#部署到好雨云)
  - [一键部署](#一键部署)
  - [配置向导](#配置向导)
    - [欢迎页面](#欢迎页面)
    - [授权协议](#授权协议)
    - [系统检查](#系统检查)
    - [配置数据库](#配置数据库)
    - [设置帐号](#设置帐号)
    - [安装完成](#安装完成)
    - [登录](#登录)
- [参与和讨论](#参与和讨论)
- [版权说明](#版权说明)

# 部署到好雨云
## 一键部署
通过点击本文最上方的 “部署到好雨云” 按钮会跳转到 好雨应用市场的应用首页中，可以通过一键部署按钮安装


**部署然之**

<img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101441582057.png" width="70%" height="70%"></img>


**注意：** 由于然之协同系统需要MySQL数据库，因此部署时会提示用户新建MySQL应用或者选择一个已有的MySQL应用。

## 配置向导
部署完成后 应用首页 点击 “访问” 按钮会跳转到然之的安装向导页面，如下图：

### 欢迎页面

<img src="http://ojfzu47n9.bkt.clouddn.com/201704011491014427287.png" width="70%" height="70%"></img>


### 授权协议

<img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101443618407.png" width="70%" height="70%"></img>

### 系统检查

<img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101444620364.png" width="70%" height="70%"></img>

### 配置数据库
> 请根据关联的MySQL实际情况填写连接信息，可以在MySQL首页，或者然之的依赖页面查看到MySQL的连接信息。

<img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101445971358.png" width="70%" height="70%"></img>


### 设置管理帐号

<img src="http://ojfzu47n9.bkt.clouddn.com/20170401149101446998912.png" width="60%" height="60%"></img>

### 安装完成

> - 安装完成后平台会自动删除 `install.php` 和 `upgrade.php` 文件
> - 如果需要重新初始化配置，只需要删除 `../config/my.php` 文件即可

### 登录

<img src="http://ojfzu47n9.bkt.clouddn.com/2017040114910144785552.png" width="50%" height="50%"></img>

**然之系统截图**

<img src="http://ojfzu47n9.bkt.clouddn.com/2017040114910144868964.png" width="60%" height="60%"></img>

# 参与和讨论
如果您对本项目感兴趣或有疑问可以在好雨讨论社区[发表评论](http://t.goodrain.com/c/11-category)

# 版权说明
本项目同步更新 [然之](https://www.ranzhico.com/download.html) 官方发布的开源版本，并适配好雨云的一键部署 功能。

[然之](https://www.ranzhico.com/) 开源版本 版权归[青岛易软天创网络科技有限公司](http://www.cnezsoft.com/)所有并遵循原软件的[版权规则](https://github.com/goodrain-apps/ranzhi/blob/master/doc/LICENSE)


doc/CHANGELOG
2017-02-27 4.1版本发布
修复通用搜索日期类的条件选择昨天，搜索后再次打开该选择会变为英文状态的1899年的bug
修复记账收入创建的时候选择合同，编辑时合同值为空的bug
修复然之集成GET访问方式的禅道，无法添加禅道的区块的bug
修复集成禅道后添加禅道的我的BUG区块，无法正常显示的bug
修复文档内容贴图报错的bug
修复crm产品列表点击表头排序会报错的bug
修复proj应用权限设置失效的bug
修复组织用户点击标题排序报错的bug
将error类改为notice避免与php7内置的冲突
修改动态里面的项目的链接，跳转到项目应用
调整cash-收入菜单的高亮

2017-01-23 4.0版本发布
将文档独立出来作为一个应用
移植禅道开源版本的文档管理的功能
项目增加文档库功能
文档库增加排序功能
将项目独立出来作为一个应用
优化调整然之各个应用的配色
合并客户功能
调整客户列表里面的时间顺序：添加，最后联系，下次联系。
调整合同和订单相关区块更多的链接
优化会员登录页面之后跳转referer的逻辑
增加附件编辑删除权限管理
记账的时候显示客户/供应商的对公账户
支出如果没有权限的话，提示用户
批量记账提供同上功能
论坛功能支持调整每页显示条数和快速跳转页面
报销列表里面的标题和科目增加title
待办在全屏状态切换的时候，表头固定要重新计算宽度

2016-10-25 3.7版本发布
优化然之的性能
优化misc-ping的性能
重构帐目明细表
重构考勤统计功能
重构账号盈亏报表
重构export2excel类
增加调休的功能
增加外出管理
实现销假的功能
我的地盘审批增加调休审批
我的地盘增加未关闭的任务列表
动态增加搜索功能
加班类型调休改为补班，超时加班改为工作日加班
加班审核拒绝的时候增加拒绝理由的填写
优化记账模块统计报表的样式
现金记账模块，在线支付类型增加微信支付
报表数据以万为单位显示
用户列表禁用和删除的账户默认不予列出
供应商（客户）增加对公账户字段
项目列表添加搜索功能
项目任务列表增加一个返回菜单
任务增加批量关闭功能
完成的待办增加激活功能
导入待办改为移动
报销的时候科目显示完整路径的科目
报销列表页提供搜索和报表功能(页脚的统计功能)
调整应用分组功能
在CRM里面增加供应商入口
创建节假日之后发送提醒

2016-10-08 3.6版本发布
升级到最新的zentaophp框架
待办列表页面批量操作增加其他操作
然之的api增加https接口
合同回款时选择关联记账的时候把回款的备注填入记账的说明字段内
科目可以调整顺序
手续费，理财盈利，理财亏损，贷款利息作为系统内置科目
将投资改为理财，记录分为投资和借贷
投资记录需要标记有没有赎回
调整报销提醒邮件的样式
报销增加默认排序
报销自动记账时将报销的经手人和明细信息也记录到记账模块
考勤应该可以设置节假日为工作日
创建客户的时候可以关联已有的联系人
优化销售分组功能
解决博客类目无法删除的问题
user表加account的唯一性检查
记账列表页进入的时候默认是当前季度的第一个月份
记账明细保存提示成功但是查询不到
解决报销明细科目第二行无法搜索中间的字符串的问题
user的getPairs方法过滤被禁止的用户时没有判断locked时间是否已过期
客户、联系人列表页根据权限过滤后右下角的总条目数还是未过滤的
直接访问OA应用的某个页面，跳转到登录地址是OA的
出差和请假同一天的时候，请假无法添加
报销审批时不填写报销金额也可以保存，结果是报销0元
草稿状态的报销无法编辑，提示信息访问受限
customer/model.php getCustomersSawByMe方法查询订单的时候in条件没有trim逗号，导致会把未指派的订单查询出来
合同操作动态没有进行销售分组权限过滤，合同详情页面没有检查销售分组权限
客户可以在指派的时候设为公共客户（进入客户池），但是却无法在指派页面里取消公共客户属性，必须到编辑页面去取消
超级管理员能看到所有人的需要审核的信息
云然之应用的集成页面里的返回按钮不生效
OA-考勤-设置-部门经理设置，空的时候返回内容错误
任务完成页面，点保存后保存成功提示位置错误。
客户报表无法根据搜索结果生成
搜索帐目时，导航里面选择2016，搜索表单搜索2015年的无结果
当应用打开方式为内嵌窗口时，打开应用显示空白。
日历区块添加待办时起止时间显示错误

2016-07-19 3.5版本发布
报销、请假、加班增加csv导出功能
增加考勤明细浏览功能
增加导出考勤明细的功能
考勤如果设置必须签退，取消自动签退功能
考勤可以设置不考勤的人员名单
修改考勤统计
加班、请假支持一天记录多次
每天的请假时长不应大于工作工时
请假、加班、出差的时间不能重复
解决连续多天请假、出差时在考勤中显示请假、出差时间错误的问题
分页增加跳转到第几页
调整记账模块的功能
优化统计报表
增加导出按照账号的收入盈亏表的功能
系统科目内置主营业务收入、非主营业务收入、主营业务成本、非主营业务成本
记账模块的搜索功能增加产品搜索，科目根据记账类型显示不同科目
客户搜索增加合作伙伴类型的筛选条件
合同按照销售分组的权限分开
我的地盘审核菜单增加权限判断
个人报销默认按照状态排序

2016-05-24 3.4版本发布
待办集成禅道的任务和bug
订单，客户，联系人等增加备注功能
优化客户联系人查询效率
公告记录已经阅读的用户
供应商增加分类管理
报销记录关联订单、合同、客户
报销设置由谁报销，加强报销的流程
报销增加金额设置，小于此金额只需要一级审批，大于此金额需要二级审批
设置默认的报销账户
增加由我审批的报销列表
报销可以按照月份进行浏览
报销审批列表增加查看详情的按钮
完善现金记账的区块，增加收入、支出类型的选择
记账里的收入，支出，转账，投资，报表放在第一级菜单
账目列表年份和月份的选择做成横向的选择
报表按照数据排序
账目列表产品和客户居左对齐
账目可以上传附件
将后台管理部分权限放入权限管理内
调整沟通页面的样式
优化任务列表页面的按钮
调整任务单个创建页面的样式
任务区块检查可以操作的菜单列表
加班增加详情页面
默认不提供地域和行业数据的安装
然之插件添加ZPL协议
修复待办的表头固定的问题
修复然之禅道集成密码有@等特殊字符时提示密码错误的问题
修复退出登录前地址是集成应用时(如禅道)，再次登录会自动跳转至应用，没有然之界面的问题
修复禅道集成专业版时提示小于7.4的问题
修复项目窗口缩放时，任务列表表格展示不对的问题
修复待办拖动的时候，切换到上个月，拖动到本月时，不显示的问题
修复看板按照指派给分组，将未指定的拖动到某个人时，没有修改指派给字段的问题
修复项目列表只在当前页排序，不是全局排序的问题

2016-04-18 3.3版本发布
左侧应用添加分组功能
考勤可以限定签到的ip地址
增加限制ip登录的功能
美化邮件提醒的模板
计划任务默认打开
备份可以设置存储的时间
科目增加是否是主营业务的属性
提供科目合并功能
浏览支出和收入记录的时候，可以组合左侧的月份浏览
账目列表改为产品/科目
待办列表可以点击一行直接选中前面的checkbox
日历待办的宽度保持一致
增加有无待办的提醒，每天提醒一次，上班后半小时提醒
优化待办创建页面样式
修复待办自己指派给自己不显示的bug
项目列表增加默认的排序
修复项目看板页面表头固定的bug
修复没有操作所有项目任务权限时，项目的成员（非创建者、负责人），完成任务时提示无权限的bug
我的地盘里面我的项目的描述控制在一行显示
我的地盘里面我的项目的创建项目的菜单放到表头
调整沟通功能的样式，分页显示沟通记录
名单忽略的时候可以写明原因
名单增加设置功能，可以设置派发规则
新建订单的产品列表里只显示状态为正常的产品
报销审批人和创建人相同的时候默认审核通过
报销审核界面样式调整
报销列表显示报销科目
报销设置页面菜单项调整
请假的审批人提供编辑功能
集成禅道的时候如果版本不符合要求，提示用户
任务搜索添加创建者、完成者条件
chosen下拉列表都居左显示
调整插件详情页面
测试ie11的兼容性问题

2016-02-24 3.2.1版本发布
解决加班、请假审核人获取需要审核的列表的问题
解决博客列表排序问题
解决没有订单时生成报表会出错的问题
不支持bccomp方法的时候给出提示
合同详情界面编辑回款时记录备注信息
解决客户沟通新建联系人时界面出错的问题
处理记账日期是每年最后一天不会被统计的问题
项目的相关操作只有项目创建人和项目负责人有权限
解决项目列表的分页问题
任务完成时指派给项目团队的人

2016-02-16 3.2版本发布
新增名单功能
考勤增加统计功能
文档库，文档，博客模块，博客文章，论坛版块增加权限控制功能
OA里面的考勤，出差，请假，报销功能设置开关
禅道集成解决https的问题
实现权限的实时生效功能
公用权限的设置增加备注
待定里点击完成操作的事务，显示在操作当天的待办日历里
未完成的待办点击完成之后，从未完成列表中移除
待办日历滚动的时候，周几的那一行固定在顶部，左侧的日期选择也固定在顶部
指派给他人的待办显示下指派人和被指派人
待办的日历周末调整显示的宽度
组织待办显示请假或者出差信息
组织根据考勤里面设置的工作日，来显示上个工作日
组织成员搜索支持真实姓名搜索
合同的回款功能自动和帐目关联
我的合同区块增加由我创建选项
项目任务默认列表是所有任务
任务列表菜单增加项目的维护菜单
查看子任务的时候显示父任务的名称
多人任务可以调整顺序
项目任务增加颜色的区分
项目的待办看板增加颜色的区别
看板的任务描述控制下显示的内容
调整项目区块
把记账的应用服务改为$
沟通的时候，选择某个联系人，可以直接把联系方式显示出来
调整日志的记录逻辑
优化订单的报表样式
二级菜单固定位置

2015-12-22 3.1版本发布
我的地盘里面增加新增对象的入口
我的地盘我的订单，我的合同的检索条件进行约束
优化贴图逻辑
待办那边增加未完成的列表
增加未完成和待定的待办可以导入到某一天的功能
组织待办可以添加备注信息
组织待办默认按照部门排序
批量添加待办的时候时候，时间也可以全部待定
修复创建待办的时候如果选择了日期，弹窗页面变全屏的问题
优化组织待办在滚屏情况下查看待办信息的样式问题
优化待办提醒
记录登录日志
合同列表和合同详情页面显示合同编号
项目列表加分页
修复项目任务无法按照状态进行排序的问题
修复左侧菜单拖动排序的问题

2015-12-02 3.0版本发布
增加我的地盘应用：待办，任务，项目，订单，合同，审批，组织，动态
增加扁平风格
实现消息提醒机制
实现版本更新提醒功能
区块增加更多链接
区块增加动态区块
梳理默认的区块布局，保证能够占满一行
添加应用操作拆分简化
增加单个任务的创建页面
任务添加备注功能
沟通记录的附件集中显示
待办增加列表页面
日历区块内的待办可以操作
客户的邮箱可以填写多个
已完成的项目增加列表显示页面
账户增加标签功能
记录帐目浏览的目录状态
供应商列表的区域和行业居左对齐
统一导入导出的图标按钮
左下角的图标鼠标悬浮增加当前登录用户名提示
右下角增加然之协同的版权标志和版本号

2015-11-13 2.7版本发布
授权协议修改为 ZPL 1.2
提供rpm、deb和Linux一键安装包
实现待办的指派功能
OA我的地盘里面增加和我相关的列表
导入帐目的时候自动去重功能
调整现金流报表的样式
与禅道集成，增加同步用户操作
增加请假，报销的撤销功能
请假报销添加邮件提醒
休息日登录然之，状态为加班
优化index页面迟到提醒
报销可以添加附件
梳理出差，请假，考勤，报销，待办权限设置
区块里面的日程可以直接添加内容
任务区块增加基本的操作按钮
联系人沟通页面提供附件上传功能
平板电脑上面沟通记录无法滚动浏览

2015-10-29 2.6版本发布
新增报销功能
新增日历功能，待办可以拖动到日历里面
项目任务可以指定给多人
项目任务支持子任务
完善项目、任务的权限控制
完善任务浏览时候上一个下一个的逻辑
简化然之和禅道集成的设置
调整考勤的补录功能
考勤和日历的待办合并
考勤模块增加部门经理的设置
考勤模块可以设置一个统一的审核人
调整考勤模块导出的excel表格模式
优化考勤页面的样式
如果没有设置必须签退，不需要提示签退
考勤设置里面增加周日开始选项

2015-08-29 2.5版本发布
在OA里面添加考勤、请假、出差功能
实现考勤参数设置功能
实现节假日增删改查
实现用户登录时间、最后更新时间和签到时间整合
实现我的考勤记录功能
实现查看我的部门的考勤记录的功能
实现查看组织的考勤记录功能
实现考勤记录的导出csv功能
实现部门负责人的维护功能
实现考勤记录的补录功能
实现考勤记录的审核功能
实现个人考勤记录日历的区块功能
在桌面右下角放置签到签退快捷按钮
实现用户的请假功能
实现请假的审核功能
实现查看我的请假记录功能
实现查看部门的请假记录功能
实现查看整个公司的请假记录功能
实现出差的添加功能
实现查看我的出差记录功能
实现查看部门的出差记录功能
实现查看所有的出差记录功能

2015-08-04 2.4版本发布
现金报表统计可以查看全年的支出收入统计
可以设置经过多少时间之后自动将没有更新的客户转入客户池
客户的联系人增加移除功能
创建订单的时候可以直接创建产品
一键安装包修改服务名称，和禅道的apachezt区分开
模块扩展时临时文件增加一个app的前缀加以区分
编辑客户的时候禁用左右方向键的切换功能
历史记录用户选择排序之后记录cookie
记录浏览记账页面左侧折叠的状态
浏览记账页面左侧折叠之后，底部按钮和表头做相应调整
把ajaxform的超时时间调大
修复kindEditor的swf上传漏洞
修复windows一键安装包ico图标
调整一键安装包首页的欢迎使用的页面布局
调整支出浏览权限设置按钮
修复“我参与的项目”不显示项目的问题
修复看板视图下的任务拖拽后，无法关闭模态框的问题
现金记账报表默认选择当前月
调整报表表格的最大高度
修复大纲视图下面操作按钮有两排的问题
修复编辑收入时合同信息丢失问题
修复订单的详情页面的历史记录在ie8下面显示不正常

config/config.php
<?php
/**
 * The config file of RanZhi.
 *
 * Don't modify this file directly, copy the item to my.php and change it.
 *
 * @copyright   Copyright 2009-2015 青岛易软天创网络科技有限公司(QingDao Nature Easy Soft Network Technology Co,LTD, www.cnezsoft.com)
 * @license     ZPL (http://zpl.pub/page/zplv12.html)
 * @author      Chunsheng Wang <chunsheng@cnezsoft.com>
 * @package     config
 * @version     $Id: config.php 4211 2016-10-25 01:38:10Z daitingting $
 * @link        http://www.ranzhi.org
 */
/* Judge class config and function getWebRoot exists or not, make sure php shells can work. */
if(!class_exists('config')){class config{}}
if(!function_exists('getWebRoot')){function getWebRoot(){}}
if(!isset($config)) $config = new stdclass();

/* 系统框架配置。Framework settings. */
$config->framework = new stdclass();
$config->framework->autoConnectDB  = true;  // 是否自动连接数据库。              Whether auto connect database or not.
$config->framework->multiLanguage  = true;  // 是否启用多语言功能。              Whether enable multi lanuage or not.
$config->framework->multiTheme     = true;  // 是否启用多风格功能。              Whether enable multi theme or not.
$config->framework->detectDevice   = true;  // 是否启用设备检测功能。            Whether enable device detect or not.
$config->framework->multiSite      = false; // 是否启用多站点模式。              Whether enable multi site mode or not.
$config->framework->extensionLevel = 1;     // 0=>无扩展,1=>公共扩展,2=>站点扩展 0=>no extension, 1=> common extension, 2=> every site has it's extension.
$config->framework->jsWithPrefix   = true;  // js::set()输出的时候是否增加前缀。 When us js::set(), add prefix or not.
$config->framework->filterBadKeys  = true;  // 是否过滤不合要求的键值。          Whether filter bad keys or not.
$config->framework->filterTrojan   = true;  // 是否过滤木马攻击代码。            Whether strip trojan code or not.
$config->framework->filterXSS      = true;  // 是否过滤XSS攻击代码。             Whether strip xss code or not.
$config->framework->purifier       = false; // 是否对数据做purifier处理。        Whether purifier data or not.
$config->framework->logDays        = 14;    // 日志文件保存的天数。              The days to save log files.

/* Basic settings. */
$config->version      = '4.2';             // The version of ranzhi. Don't change it.
$config->charset      = 'UTF-8';           // The charset of ranzhi.
$config->cookieLife   = time() + 2592000;  // The cookie life time.
$config->timezone     = 'Asia/Shanghai';   // The time zone setting, for more see http://www.php.net/manual/en/timezones.php
$config->cookiePath   = '/';               // The path of cookies.
$config->webRoot      = getWebRoot();      // The web root.
$config->checkVersion = true;              // Auto check for new version or not.
$config->timeout      = 30 * 1000;         // The timeout of ajax request.
$config->pingInterval = 60;                // The interval of ping request, seconds.

/* The request settings. */
$config->requestType = 'PATH_INFO';        // The request type: PATH_INFO|GET, if PATH_INFO, must use url rewrite.
$config->requestFix  = '-';                // The divider in the url when PATH_INFO.
$config->moduleVar   = 'm';                // requestType=GET: the module var name.
$config->methodVar   = 'f';                // requestType=GET: the method var name.
$config->viewVar     = 't';                // requestType=GET: the view var name.
$config->sessionVar  = 'rid';              // requestType=GET: the session var name.

/* Supported views. */
$config->views = ',html,json,mhtml,'; 

/* Supported themes. */
$config->themes['default'] = 'default'; 

/* Supported languages. */
$config->langs['zh-cn'] = '简体';
$config->langs['zh-tw'] = '繁体';
$config->langs['en']    = 'English';

/* The prefix for view file for different device. */ 
$config->devicePrefix['mhtml'] = 'm.';

/* Supported charsets. */
$config->charsets['zh-cn']['utf-8'] = 'UTF-8';
$config->charsets['zh-cn']['gbk']   = 'GBK';
$config->charsets['zh-tw']['utf-8'] = 'UTF-8';
$config->charsets['zh-tw']['big5']  = 'BIG5';
$config->charsets['en']['utf-8']    = 'UTF-8';

/* Default settings. */
$config->default = new stdclass();
$config->default->view   = 'html';        // Default view.
$config->default->lang   = 'en';          // Default language.
$config->default->theme  = 'default';     // Default theme.
$config->default->module = 'index';       // Default module.
$config->default->method = 'index';       // Default method.

$config->dashboard = new stdclass();
$config->dashboard->modules = 'my,todo';

/* Upload settings: danger files and max upload size. */
$config->file = new stdclass();
$config->file->dangers = 'php,php3,php4,phtml,php5,jsp,py,rb,asp,aspx,ashx,asa,cer,cdx,aspl,shtm,shtml,html,htm';
$config->file->maxSize = 1024 * 1024;

/* IP white list settings.*/
$config->ipWhiteList = '*';

/* Set the allowed tags.  */
$config->allowedTags = '<p><span><h1><h2><h3><h4><h5><em><u><strong><br><ol><ul><li><img><a><b><font><hr><pre><div><table><td><th><tr><tbody>';

/* Master database settings. */
$config->db = new stdclass();
$config->db->persistant     = false;     // Pconnect or not.
$config->db->driver         = 'mysql';   // Must be MySQL. Don't support other database server yet.
$config->db->encoding       = 'UTF8';    // Encoding of database.
$config->db->strictMode     = false;     // Turn off the strict mode of MySQL.
//$config->db->emulatePrepare = true;    // PDO::ATTR_EMULATE_PREPARES
//$config->db->bufferQuery    = true;    // PDO::MYSQL_ATTR_USE_BUFFERED_QUERY

/* Slave database settings. */
$config->slaveDB = new stdclass();
$config->slaveDB->persistant  = false;      
$config->slaveDB->driver      = 'mysql';    
$config->slaveDB->encoding    = 'UTF8';     
$config->slaveDB->strictMode  = false;      

/* Include the custom config file. */
$configRoot = dirname(__FILE__) . DIRECTORY_SEPARATOR;
$myConfig   = $configRoot . 'my.php';
if(file_exists($myConfig)) include $myConfig;
$rightsConfig = $configRoot . 'rights.php';
if(file_exists($rightsConfig)) include $rightsConfig;

/* Tables for basic system. */
define('TABLE_CONFIG',    '`sys_config`');
define('TABLE_PACKAGE',   '`sys_package`');
define('TABLE_USER',      '`sys_user`');
define('TABLE_GROUP',     '`sys_group`');
define('TABLE_ACTION',    '`sys_action`');
define('TABLE_FILE',      '`sys_file`');
define('TABLE_HISTORY',   '`sys_history`');
define('TABLE_CATEGORY',  '`sys_category`');
define('TABLE_ARTICLE',   '`sys_article`');
define('TABLE_EXTENSION', '`sys_extension`');
define('TABLE_WEBAPP',    '`sys_webapp`');
define('TABLE_LANG',      '`sys_lang`');
define('TABLE_ENTRY',     '`sys_entry`');
define('TABLE_SSO',       '`sys_sso`');
define('TABLE_TASK',      '`sys_task`');
define('TABLE_TEAM',      '`sys_team`');
define('TABLE_TAG',       '`sys_tag`');
define('TABLE_BLOCK',     '`sys_block`');
define('TABLE_SCHEMA',    '`sys_schema`');
define('TABLE_RELATION',  '`sys_relation`');
define('TABLE_CRON',      '`sys_cron`');
define('TABLE_USERGROUP', '`sys_usergroup`');
define('TABLE_GROUPPRIV', '`sys_grouppriv`');
define('TABLE_USERQUERY', '`sys_userquery`');

/* Tables for crm. */
define('TABLE_ADDRESS',       '`crm_address`');
define('TABLE_PRODUCT',       '`sys_product`');
define('TABLE_ORDER',         '`crm_order`');
define('TABLE_CUSTOMER',      '`crm_customer`');
define('TABLE_RESUME',        '`crm_resume`');
define('TABLE_CONTACT',       '`crm_contact`');
define('TABLE_CONTRACT',      '`crm_contract`');
define('TABLE_CONTRACTORDER', '`crm_contractorder`');
define('TABLE_PLAN',          '`crm_plan`');
define('TABLE_DELIVERY',      '`crm_delivery`');
define('TABLE_SALESGROUP',    '`crm_salesgroup`');
define('TABLE_SALESPRIV',     '`crm_salespriv`');

/* Tables for oa. */
define('TABLE_TODO',       '`oa_todo`');
define('TABLE_PROJECT',    '`oa_project`');
define('TABLE_DOC',        '`oa_doc`');
define('TABLE_DOCCONTENT', '`oa_doccontent`');
define('TABLE_DOCLIB',     '`oa_doclib`');
define('TABLE_ATTEND',     '`oa_attend`');
define('TABLE_ATTENDSTAT', '`oa_attendstat`');
define('TABLE_HOLIDAY',    '`oa_holiday`');
define('TABLE_LEAVE',      '`oa_leave`');
define('TABLE_OVERTIME',   '`oa_overtime`');
define('TABLE_LIEU',       '`oa_lieu`');
define('TABLE_TRIP',       '`oa_trip`');
define('TABLE_REFUND',     '`oa_refund`');

/* Tables for cash. */
define('TABLE_DEPOSITOR', '`cash_depositor`');
define('TABLE_BALANCE',   '`cash_balance`');
define('TABLE_TRADE',     '`cash_trade`');

/* Tables for team. */
define('TABLE_THREAD',  '`team_thread`');
define('TABLE_REPLY',   '`team_reply`');
define('TABLE_MESSAGE', '`sys_message`');

/* The mapping list of object and tables. */
$config->objectTables['announce']    = TABLE_ARTICLE;
$config->objectTables['article']     = TABLE_ARTICLE;
$config->objectTables['attend']      = TABLE_ATTEND;
$config->objectTables['blog']        = TABLE_ARTICLE;
$config->objectTables['contact']     = TABLE_CONTACT;
$config->objectTables['contract']    = TABLE_CONTRACT;
$config->objectTables['cron']        = TABLE_CRON;
$config->objectTables['customer']    = TABLE_CUSTOMER;
$config->objectTables['depositor']   = TABLE_DEPOSITOR;
$config->objectTables['doc']         = TABLE_DOC;
$config->objectTables['doclib']      = TABLE_DOCLIB;
$config->objectTables['holiday']     = TABLE_HOLIDAY;
$config->objectTables['leads']       = TABLE_CONTACT;
$config->objectTables['leave']       = TABLE_LEAVE;
$config->objectTables['lieu']        = TABLE_LIEU;
$config->objectTables['makeup']      = TABLE_OVERTIME;
$config->objectTables['order']       = TABLE_ORDER;
$config->objectTables['overtime']    = TABLE_OVERTIME;
$config->objectTables['product']     = TABLE_PRODUCT;
$config->objectTables['project']     = TABLE_PROJECT;
$config->objectTables['provider']    = TABLE_CUSTOMER;
$config->objectTables['refund']      = TABLE_REFUND;
$config->objectTables['reply']       = TABLE_REPLY;
$config->objectTables['resume']      = TABLE_RESUME;
$config->objectTables['schema']      = TABLE_SCHEMA;
$config->objectTables['task']        = TABLE_TASK;
$config->objectTables['thread']      = TABLE_THREAD;
$config->objectTables['todo']        = TABLE_TODO;
$config->objectTables['trade']       = TABLE_TRADE;
$config->objectTables['user']        = TABLE_USER;

/* Include extension config files. */
$extConfigFiles = glob($configRoot . 'ext/*.php');
if($extConfigFiles) foreach($extConfigFiles as $extConfigFile) include $extConfigFile;
